From 0afc1f0df8e150c1a243fa3cadce0ae5ed425e11 Mon Sep 17 00:00:00 2001
From: Sean Harmer <sean.harmer@kdab.com>
Date: Wed, 21 Oct 2020 12:11:39 +0100
Subject: [PATCH 11/18] Trigger an update on the quick window when creating a
 render aspect

Sometimes if a scene is large there is a timing issue that prevents
Qt Quick from knowing it needs to do an update.

When creating a new RenderAspect give the QQuickWindow a nudge to
force an update.

Change-Id: Id0bd597ac8845ec5d9c89c7aa00ad57ae6b0056d
Reviewed-by: Nicolas Arnaud-Cormos <nicolas@kdab.com>
Reviewed-by: Mike Krus <mike.krus@kdab.com>
(cherry picked from commit df63c58ce6872d5b0c62bf34763ac7c093437a81)
Reviewed-by: Paul Lemire <paul.lemire@kdab.com>
---
 src/quick3d/imports/scene3d/scene3ditem.cpp | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/src/quick3d/imports/scene3d/scene3ditem.cpp b/src/quick3d/imports/scene3d/scene3ditem.cpp
index 5c78d37ac..126680595 100644
--- a/src/quick3d/imports/scene3d/scene3ditem.cpp
+++ b/src/quick3d/imports/scene3d/scene3ditem.cpp
@@ -925,6 +925,8 @@ QSGNode *Scene3DItem::updatePaintNode(QSGNode *node, QQuickItem::UpdatePaintNode
                     m_framesToRender = ms_framesNeededToFlushPipeline;
                 },
                 Qt::DirectConnection);
+        // Give the window a nudge to trigger an update.
+        QMetaObject::invokeMethod(window(), "requestUpdate", Qt::QueuedConnection);
     }
 
     const bool usesFBO = m_compositingMode == FBO;
-- 
2.35.1

